### PR TITLE
drivers: rtc: new maxim ds1307 rtc driver

### PR DESCRIPTION
**Pull request by arun-mani-tech**
Added rtc driver for ds1307. It allows to read and set the date and time

### CODE REVIEW

**Comment by arun-mani-tech** [comment]
@bjarki-trackunit Kindly have a look

File: drivers/rtc/rtc_ds1307.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+#define MONTHS_BITS  GENMASK(4, 0)
+#define WEEKDAY_BITS GENMASK(2, 0)
+#define YEAR_BITS    GENMASK(7, 0)
+
+struct ds1307_config {
+	struct i2c_dt_spec i2c_bus;
+};
+
+struct ds1307_data {
+	struct k_mutex lock;
```
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_ds1307.c:44]: You should use a spinlock here, since mutex ownership is not used :)

File: drivers/rtc/rtc_ds1307.c [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+	}
+
+	timeptr->tm_sec = bcd2bin(regs[0] & SECONDS_BITS);
+	timeptr->tm_min = bcd2bin(regs[1] & MINUTES_BITS);
+	timeptr->tm_hour = bcd2bin(regs[2] & HOURS_BITS); /* 24hr mode */
+	timeptr->tm_wday = bcd2bin(regs[3] & WEEKDAY_BITS);
+	timeptr->tm_mday = bcd2bin(regs[4] & DATE_BITS);
+	timeptr->tm_mon = bcd2bin(regs[5] & MONTHS_BITS);
+	timeptr->tm_year = bcd2bin(regs[6] & YEAR_BITS);
+
```
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_ds1307.c:99]: You should add a check here to validate the chip is indeed in 24 hour mode, if not, return -ENODATA as the time has not been configured correctly :)

File: drivers/rtc/rtc_ds1307.c [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+	tx_buf[0] = DS1307_REG_SECONDS;
+	tx_buf[1] = 0x00;
+
+	/* Make clock halt = 0 */
+	err = i2c_write_read_dt(&config->i2c_bus, tx_buf, sizeof(tx_buf), &ch, sizeof(ch));
+	if (err < 0) {
+		LOG_ERR("Error: Set clock halt bit:%d\n", err);
+	}
+
+	return ((ch >> 7) & 0x1);
```
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_ds1307.c:144]: The init function must succeed if the device was initialized properly, not if its configured properly (time is set). If the time is not set, the get_time function will fail with -ENODATA with the addition i suggested, and the device will be configured once the time is set on the next rtc_set_time().

A failing init will prevent the system from booting, preventing it from setting the time correctly later

File: drivers/rtc/rtc_ds1307.c [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+static const struct rtc_driver_api ds1307_driver_api = {
+	.set_time = ds1307_set_time,
+	.get_time = ds1307_get_time,
+};
+
+/*
+ * returns 1 : CH = 1; init failed
+ * returns 0 : CH = 0; init success
+ */
```
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_ds1307.c:119]: Remove these comments based on the suggestion later in the init function :)

File: drivers/rtc/rtc_ds1307.c [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	const struct ds1307_config *config = dev->config;
+	struct ds1307_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+
+	if (!i2c_is_ready_dt(&config->i2c_bus)) {
+		LOG_ERR("I2C bus not ready");
+		return -ENODEV;
+	}
+
```
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_ds1307.c:134]: Add an init of the control register here, ensuring no square wave is sent to any output. You can add configuration of the square wave to the devicetree if you want to, but without it, default to no square wave :)

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
Overall it looks good, I have a few requests I have outlined that need to be addressed :)

**Comment by arun-mani-tech** [comment]
@bjarki-trackunit Please review

**Comment by bjarki-trackunit** [comment]
Looks good, last request, could you run the [RTC API test suite](https://docs.zephyrproject.org/latest/hardware/peripherals/rtc.html#rtc-device-driver-test-suite)? (ignore the y2k test failing once you run it)

**Comment by arun-mani-tech** [comment]
@bjarki-trackunit I ran RTC API test_suite. All checks are passed except yday, isdst, nsec. Because this chip does not support that

File: drivers/rtc/rtc_ds1307.c [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		goto unlock;
+	}
+
+	timeptr->tm_sec = bcd2bin(regs[0] & SECONDS_BITS);
+	timeptr->tm_min = bcd2bin(regs[1] & MINUTES_BITS);
+	timeptr->tm_hour = bcd2bin(regs[2] & HOURS_BITS); /* 24hr mode */
+	timeptr->tm_wday = bcd2bin(regs[3] & WEEKDAY_BITS);
+	timeptr->tm_mday = bcd2bin(regs[4] & DATE_BITS);
+	timeptr->tm_mon = bcd2bin(regs[5] & MONTHS_BITS);
+	timeptr->tm_year = bcd2bin(regs[6] & YEAR_BITS);
```
Comment (bjarki-trackunit) [inline drivers/rtc/rtc_ds1307.c:98]: You must initialize all fields of the `struct tm` according to the documentation here https://docs.zephyrproject.org/latest/hardware/peripherals/rtc.html#c.rtc_time.tm_yday

So add
```
	timeptr->tm_yday = -1;
	timeptr->tm_isdst =  -1;
	timeptr->tm_nsec =  0;
```

**Review by bjarki-andreasen — CHANGES_REQUESTED** [review]
This last change should allow it to fail only because of the missing yday :)

File: drivers/rtc/rtc_ds1307.c (hunk 6 above)
Comment (bjarki-trackunit) [reply drivers/rtc/rtc_ds1307.c:98]: I'm fixing the test as well with this PR https://github.com/zephyrproject-rtos/zephyr/pull/64447, you don't have to wait for it to be merged or anything, just pointing it out :)

**Review by bjarki-andreasen — DISMISSED** [review]
Nice work!

**Comment by arun-mani-tech** [comment]
@bjarki-trackunit Thanks for the review

File: drivers/rtc/rtc_ds1307.c [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+	timeptr->tm_year = timeptr->tm_year + 100;
+
+	/* Not used */
+	timeptr->tm_nsec =  0;
+	timeptr->tm_isdst =  -1;
+	timeptr->tm_yday = -1;
+
+	/* Validate the chip in 24hr mode */
+	if (regs[2] & (1 << 6)) {
+		return -ENODATA;
```
Comment (fabiobaltieri) [inline drivers/rtc/rtc_ds1307.c:108]: this should probably set `err` and `goto unlock;`

File: drivers/rtc/rtc_ds1307.c [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+	/* Make clock halt = 0 */
+	err = i2c_write_dt(&config->i2c_bus, tx_buf, sizeof(tx_buf));
+	if (err < 0) {
+		LOG_ERR("Error: Set clock halt bit:%d\n", err);
+	}
+
+	return 0;
+}
+
+#define DS1307_INIT(inst)                                                                          \
```
Comment (fabiobaltieri) [inline drivers/rtc/rtc_ds1307.c:158]: nit: these are normally named something like `DS1307_DEFINE`, init is the initialization function, this is defining a bunch of stuff.

File: drivers/rtc/rtc_ds1307.c [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+	timeptr->tm_year = bcd2bin(regs[6] & YEAR_BITS);
+	timeptr->tm_year = timeptr->tm_year + 100;
+
+	/* Not used */
+	timeptr->tm_nsec =  0;
+	timeptr->tm_isdst =  -1;
+	timeptr->tm_yday = -1;
+
+	/* Validate the chip in 24hr mode */
+	if (regs[2] & (1 << 6)) {
```
Comment (fabiobaltieri) [inline drivers/rtc/rtc_ds1307.c:107]: you could probably use a define for this, and use `BIT(6)` to define it.

File: drivers/rtc/rtc_ds1307.c [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+	timeptr->tm_hour = bcd2bin(regs[2] & HOURS_BITS); /* 24hr mode */
+	timeptr->tm_wday = bcd2bin(regs[3] & WEEKDAY_BITS);
+	timeptr->tm_mday = bcd2bin(regs[4] & DATE_BITS);
+	timeptr->tm_mon = bcd2bin(regs[5] & MONTHS_BITS);
+	timeptr->tm_year = bcd2bin(regs[6] & YEAR_BITS);
+	timeptr->tm_year = timeptr->tm_year + 100;
+
+	/* Not used */
+	timeptr->tm_nsec =  0;
+	timeptr->tm_isdst =  -1;
```
Comment (fabiobaltieri) [inline drivers/rtc/rtc_ds1307.c:103]: nit: double space

File: drivers/rtc/rtc_ds1307.c [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+#include <zephyr/drivers/rtc.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/sys/util.h>
+
+#define DT_DRV_COMPAT maxim_ds1307
+
+LOG_MODULE_REGISTER(ds1307, CONFIG_RTC_LOG_LEVEL);
+
+#define DS1307_I2C_ADDR 0x68
```
Comment (fabiobaltieri) [inline drivers/rtc/rtc_ds1307.c:19]: leftover definition? the address is taken from the tree

File: drivers/rtc/rtc_ds1307.c [hunk 12]
```diff
@@ -0,0 +1,167 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2023 Arunmani Alagarsamy
+ * Author: Arunmani Alagarsamy  <arunmani27100@gmail.com>
+ */
+
+#include <zephyr/drivers/gpio.h>
```
Comment (fabiobaltieri) [inline drivers/rtc/rtc_ds1307.c:8]: I think you can drop this, you don't do any gpio operation in this file

File: dts/bindings/rtc/maxim,ds1307.yaml [hunk 13]
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) 2023 Arunmani Alagarsamy
+# Author: Arunmani Alagarsamy <arunmani27100@gmail.com>
+
+description: MAXIM DS1307 RTC
```
Comment (fabiobaltieri) [inline dts/bindings/rtc/maxim,ds1307.yaml:5]: nit: Maxim

File: drivers/rtc/rtc_ds1307.c [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+	timeptr->tm_year = bcd2bin(regs[6] & YEAR_BITS);
+	timeptr->tm_year = timeptr->tm_year + 100;
+
+	/* Not used */
+	timeptr->tm_nsec = 0;
+	timeptr->tm_isdst = -1;
+	timeptr->tm_yday = -1;
+
+	/* Validate the chip in 24hr mode */
+	if (regs[2] & BIT(6)) {
```
Comment (fabiobaltieri) [inline drivers/rtc/rtc_ds1307.c:104]: ok much better, can you add a `#define` for `BIT(6)` though?

File: drivers/rtc/rtc_ds1307.c [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+	if (!i2c_is_ready_dt(&config->i2c_bus)) {
+		LOG_ERR("I2C bus not ready");
+		return -ENODEV;
+	}
+
+	tx_buf[0] = DS1307_REG_CTRL;
+	tx_buf[1] = 0x00;
+
+	/* Disable squarewave output */
+	err = i2c_write_dt(&config->i2c_bus, tx_buf, sizeof(tx_buf));
```
Comment (fabiobaltieri) [inline drivers/rtc/rtc_ds1307.c:140]: I think you could use `i2c_reg_write_byte_dt(&config->i2c_bus, DS1307_REG_CTRL, 0x00);` here and drop the local buffer. A bit more compact.

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by arun-mani-tech** [comment]
@bjarki-trackunit @fabiobaltieri Gentle ping!

**Review by bjarki-andreasen — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
> @bjarki-trackunit @fabiobaltieri Gentle ping!

No need to ping, GitHub notifications already do it for you.

**Pull request merged**